### ISSUE TITLE
Alacritty opening twice Linux/i3

### ISSUE THREAD

**Issue by oarmstrong** [jwilm/alacritty#1859]
Running ArchLinux with X11, i3. I have a keybind for launching alacritty (`bindsym $smod+Return exec --no-startup-id alacritty`). Approximately one in 5 times I use this, I get two alacritty windows instead of just one.

I have confirmed this using two separate keyboards and I haven't noticed the return key sending two characters in any other applications - my first thought was that i3 was just being told to launch it twice. I have also switched back to termite and confirmed over 20 launches that the issue does not occur.

Some interesting notes about this is that there is a definite ~200ms delay before the first window opens and then the second opens very quickly after. It also appears to happen more frequently when alacritty is starting "cold" and I haven't opened a new window in a while (although I do have other alacritty windows open), but is not 100% reproducible.

Most interestingly, when I open alacritty using a terminal and not the i3 bind, I never see two windows open. I do, however, see an extra blank line in the output (and my terminal "skips" a command).

My hypothesis is this blank line (that my terminal is interpreting and running a blank command once alacritty closes) is interpreted "oddly" by i3 and causes the second instance to be launched.

Example output when I believe the problems is being triggered (note the terminal has behaved as if I hit return while alacritty is running and once closed it has skipped a prompt line):

```
~> alacritty -v
Welcome to Alacritty.
Configuration loaded from /home/ollie/.config/alacritty/alacritty.yml

device_pixel_ratio: 1
width: 1920, height: 1042
Initializing glyph cache
Finished initializing glyph cache in 0.010278526
width: 644, height: 388
Cell Size: (8 x 16)
Padding: (2 x 2)
PTY Dimensions: Line(24) x Column(80)
Initialisation complete
Font size changed: Size(20) [DPR: 1]
width: 1920, height: 1042
Goodbye.
~>
~>
```

When it runs normally, about 4/5 of the time, there is no blank line in the output and no prompt line skipped. That blank line is always in the exact same location.

This is a tricky one to explain, so hopefully it's understandable. I can provide as much further info and explanation as possible. This occurs with the config that alacritty generates when nothing exists and I'm currently on the latest version v0.2.3.

**Comment by chrisduerr** [comment]
I'm using i3 myself and know many other that do use i3 and I've never experienced anything like this before.

I'm also certain that the blank line issue you're talking about is completely unrelated. This is due to shell startup.

If it's not reproducible when run through the CLI, running it through i3 should not make any difference. Alacritty has nothing to do with keybindings you assign it to in i3, if anything that would be an i3 issue. If you could reproduce it by starting Alacritty from the CLI, it would be possible that this is an Alacritty issue, but like this all evidence seems to indicate otherwise.

My best bet would be that you still have the `i3-sensible-terminal` default binding in your config or something and you're triggering that with the same keybind.

If you have any indication that this might be an Alacritty issue, please let me know and I will re-open this issue. But with the evidence provided here, that simply doesn't seem possible.

**Comment by oarmstrong** [comment]
Thanks for the input. I hadn't considered this might be shell related, so that'll be my next line of enquiry. I have tried with `i3-sensible-terminal` and had the same results.

Appreciate your time, I'm not particularly knowledgable when it comes to terminal emulator and shell start up.

**Comment by oarmstrong** [comment]
@chrisduerr I have some more information that may be interesting. I started with a fresh Arch install with i3 and slowly added in my configuration until I could reproduce the problem.

The problem occurs as soon as I set my keyboard rate repeating settings. I usually go with `xset r rate 220 25` (the default being `600 30`) and I start reproducing the problem when I get as low as `250 30` (the second parameter makes no difference).

Now this sounds a lot like user error, but I've never experienced this with any other terminal (or any other program that I've got key bound) nor do I ever hit the key rate delay without meaning to in other usage.

I definitely don't know much about how X11 applications start up, but here's a theoretical. As I noted before, this only occurs about 1/5 - 1/10 times. When it occurs I also notice a ~200ms delay in alacritty starting. Could it be that something in the delay of alactritty starting up causing X to not realise the key has been lifted and so hit the key repeat delay, thus triggering another alacritty window to open (or the extra new line when launching from the terminal)? I imagine it's something much lower down in the stack that deals with that and alacritty would have no impact on it, so I'm probably completely wrong.

It's a stretch, I know. I'm still not insinuating it's a bug with alacritty, but it's such an odd issue that I've not seen with any other application while running that key repeat configuration before. I'd love to hear any further thoughts, it might give me an idea of how to continue down this rabbit hole.

**Comment by chrisduerr** [comment]
@oarmstrong That sounds very interesting. I'd imagine certain key repeat rates are just too slow to give you a chance to lift your finger. I'm not sure how Alacritty could affect that though, since it only reads key input, it doesn't write it.

Maybe certain programs spawn faster than others, allowing i3 to launch the program twice in a short period of time? I'm not sure how i3 interacts with blocking programs during launch, even though it should be irrelevant if it's blocking or not, it might not be possible to launch multiple things at the exact same time (even though I'm not sure why that would be the case).

**Comment by ti-mo** [comment]
This is also happening again for me after upgrading Alacritty with Glutin 0.18. During the upgrade, I got a warning that `scale_with_dpi` introduced in https://github.com/jwilm/alacritty/pull/1355 was deprecated.

I disabled this DPI scaling setting because winit would shell out to `xrandr` in the background, which takes a good 350ms to run on my machine. This was making terminal startup time annoyingly slow. Interestingly, this also killed the double-startup bug for me. My keyboard repeat rate is also fast, but not fast enough to accidentally repeat keys often, not in this case at least.

My observation is that opening two terminals at once (during this 350ms startup time) using my i3 keybind will usually spawn 3 or 4 instances. This is probably not a bug in Alacritty, but I would like to be able to force-disable winit shelling out to `xrandr`. Launching using the 'newer' `WINIT_HIDPI_FACTOR=1.0` still gives a noticeable startup delay.

@chrisduerr Any idea how this can be worked around currently?

**Comment by chrisduerr** [comment]
@ti-mo I believe the startup delay is already tracked in the winit issue tracker. They seem to be aware that querying for the display information on X11 can take a significant amount of time, so hopefully that will be fixed upstream.

https://github.com/tomaka/winit/issues/682

**Comment by rinfiyks** [comment]
This is happening to me too. When I open alacritty via an i3 shortcut (`$mod+t`), it opens two. Worth noting that it doesn't open them both in parallel. After the first one opens (which takes maybe half a second) then the second one opens after another half second.

When I open it via the terminal (any terminal - I have tried with alacritty, termite, xterm) it doesn't open two, but you see an extra new line, as if `Return` has been pressed twice rather than once (the second `Return` happens after alacritty opens).

I have noticed that if I press the shortcut `$mod+t` very quickly (i.e. the duration that the `t` key is pressed down is as short as possible) then I don't get two alacritties.
And similarly, if I press `Return` quick enough when opening from a terminal, I don't see the extra new line.

I can also get other keys to duplicate: running `sleep 0.1; alacritty` and then quickly typing the letter `a` results in `aa` in the first terminal if you time it correctly. Similar to the other delays, the first `a` is typed immediately, but the second `a` appears as soon as alacritty has finished opening.

My keyboard rate is `xset r rate 300 30` which is a little faster than default.
